File: client/test/primitives_fixture_test.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:datawallet/src/crypto/argon2id.dart';
import 'package:datawallet/src/crypto/canonical_cbor.dart';
import 'package:datawallet/src/crypto/crypto_exception.dart';
import 'package:datawallet/src/crypto/ed25519.dart';
import 'package:datawallet/src/crypto/fingerprint.dart';
import 'package:datawallet/src/crypto/random.dart';
import 'package:datawallet/src/crypto/sealed_box.dart';
import 'package:datawallet/src/crypto/secret_box.dart';
import 'package:datawallet/src/crypto/sha256.dart';
import 'package:datawallet/src/crypto/uuid_v7.dart';
import 'package:datawallet/src/crypto/x25519.dart';
import 'package:datawallet/src/fixtures/fixtures_path.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:sodium/sodium_sumo.dart';

void main() {
  late SodiumSumo sodium;
  late Directory fixturesDir;

  setUpAll(() async {
    sodium = await SodiumSumoInit.init();
    fixturesDir = resolveFixturesDir();
  });

  // ---------------------------------------------------------------------------
  // KEK derivation
  // ---------------------------------------------------------------------------

  group('KekDerivation', () {
    test('all password fixtures produce expected KEK', () {
      final raw = File('${fixturesDir.path}/inputs/passwords.expected.json')
          .readAsStringSync();
      final rows = jsonDecode(raw) as List<dynamic>;

      for (final row in rows) {
        final name = row['name'] as String;
        final password = row['password'] as String;
        final salt = _fromHex(row['kdf_salt_hex'] as String);
        final params = row['kdf_params'] as Map<String, dynamic>;
        final m = params['m'] as int;
        final t = params['t'] as int;
        final p = params['p'] as int;
        final expectedKek = row['expected_kek_hex'] as String;

        final kek = Argon2id.deriveKek(sodium, password, salt, m, t, p);

        expect(_toHex(kek), equals(expectedKek),
            reason: 'KEK for $name');
      }
    });

    test('rejects parallelism != 1', () {
      final salt = Uint8List(16);
      expect(
        () => Argon2id.deriveKek(sodium, 'test', salt, 65536, 1, 2),
        throwsA(isA<CryptoException>()),
      );
    });
  });

  // ---------------------------------------------------------------------------
  // SecretBox round-trip
  // ---------------------------------------------------------------------------

  group('SecretBoxRoundTrip', () {
    test('encrypt and decrypt round-trip', () {
      final key = Rand.bytes(sodium, 32);
      final nonce = Rand.bytes(sodium, 24);
      final plaintext = utf8.encode('hello secretbox') as Uint8List;

      final ciphertext = SecretBoxCrypto.seal(sodium, plaintext, nonce, key);
      final recovered = SecretBoxCrypto.open(sodium, ciphertext, nonce, key);

      expect(recovered, equals(plaintext));
    });

    test('wrong key fails to decrypt', () {
      final key = Rand.bytes(sodium, 32);
      final wrongKey = Rand.bytes(sodium, 32);
      final nonce = Rand.bytes(sodium, 24);
      final plaintext = utf8.encode('test') as Uint8List;

      final ciphertext = SecretBoxCrypto.seal(sodium, plaintext, nonce, key);

      expect(
        () => SecretBoxCrypto.open(sodium, ciphertext, nonce, wrongKey),
        throwsA(isA<CryptoException>()),
      );
    });
  });

  // ---------------------------------------------------------------------------
  // SealedBox
  // ---------------------------------------------------------------------------

  group('SealedBoxOpen', () {
    test('seal and open round-trip', () {
      final keypairsJson = _readJson('${fixturesDir.path}/inputs/keypairs.json');
      final seed = _fromHex(keypairsJson['alice_enc']['seed_hex'] as String);
      final kp = X25519.seedKeyPair(sodium, seed);

      final plaintext = utf8.encode('sealed box test') as Uint8List;
      final ciphertext = SealedBox.seal(sodium, plaintext, kp.publicKey);
      final recovered =
          SealedBox.open(sodium, ciphertext, kp.publicKey, kp.secretKey);

      expect(recovered, equals(plaintext));
    });

    test('open wrapped data key from fixture', () {
      final keypairsJson = _readJson('${fixturesDir.path}/inputs/keypairs.json');
      final aliceSeed = _fromHex(keypairsJson['alice_enc']['seed_hex'] as String);
      final aliceKp = X25519.seedKeyPair(sodium, aliceSeed);

      final metaJson =
          _readJson('${fixturesDir.path}/envelopes/envelope_meta.json');
      final expectedDataKey =
          _fromHex(metaJson['basic-1-recipient']['data_key_hex'] as String);

      final cbor = CanonicalCborMapper();
      final envelopeBytes = File(
              '${fixturesDir.path}/envelopes/basic-1-recipient.cbor')
          .readAsBytesSync();
      final envelope = cbor.readValue(envelopeBytes);
      final wrappings = envelope['recipient_wrappings'] as List<dynamic>;
      final wrappedDataKey =
          (wrappings.first as Map<String, dynamic>)['wrapped_data_key']
              as Uint8List;

      final dataKey =
          SealedBox.open(sodium, wrappedDataKey, aliceKp.publicKey, aliceKp.secretKey);
      expect(dataKey, equals(expectedDataKey));
    });
  });

  // ---------------------------------------------------------------------------
  // Ed25519 determinism
  // ---------------------------------------------------------------------------

  group('Ed25519Determinism', () {
    test('seedKeyPair matches expected Ed25519 public keys', () {
      final keypairsIn = _readJson('${fixturesDir.path}/inputs/keypairs.json');
      final keypairsExp =
          _readJson('${fixturesDir.path}/inputs/keypairs.expected.json');

      for (final name in keypairsIn.keys) {
        final entry = keypairsIn[name] as Map<String, dynamic>;
        if (entry['type'] != 'ed25519') continue;

        final seed = _fromHex(entry['seed_hex'] as String);
        final kp = Ed25519.seedKeyPair(sodium, seed);
        final expectedPub =
            keypairsExp[name]['public_key_hex'] as String;

        expect(_toHex(kp.publicKey), equals(expectedPub),
            reason: 'Ed25519 public key for $name');
      }
    });

    test('seedKeyPair matches expected X25519 public keys', () {
      final keypairsIn = _readJson('${fixturesDir.path}/inputs/keypairs.json');
      final keypairsExp =
          _readJson('${fixturesDir.path}/inputs/keypairs.expected.json');

      for (final name in keypairsIn.keys) {
        final entry = keypairsIn[name] as Map<String, dynamic>;
        if (entry['type'] != 'x25519') continue;

        final seed = _fromHex(entry['seed_hex'] as String);
        final kp = X25519.seedKeyPair(sodium, seed);
        final expectedPub =
            keypairsExp[name]['public_key_hex'] as String;

        expect(_toHex(kp.publicKey), equals(expectedPub),
            reason: 'X25519 public key for $name');
      }
    });

    test('auth signature matches fixture', () {
      final authJson =
          _readJson('${fixturesDir.path}/auth/nonce-and-signature.json');
      final keypairsIn = _readJson('${fixturesDir.path}/inputs/keypairs.json');

      final signerName = authJson['signer'] as String;
      final seed =
          _fromHex(keypairsIn[signerName]['seed_hex'] as String);
      final kp = Ed25519.seedKeyPair(sodium, seed);

      final nonce = _fromHex(authJson['nonce_hex'] as String);
      final prefix = _fromHex(authJson['prefix_hex'] as String);
      final expectedSignedBytesHex =
          authJson['expected_signed_bytes_hex'] as String;
      final expectedSigHex = authJson['expected_sig_hex'] as String;

      final signedBytes =
          Uint8List(prefix.length + nonce.length)
            ..setRange(0, prefix.length, prefix)
            ..setRange(prefix.length, prefix.length + nonce.length, nonce);

      expect(_toHex(signedBytes), equals(expectedSignedBytesHex));

      final sig = Ed25519.signDetached(sodium, kp.secretKey, signedBytes);
      expect(_toHex(sig), equals(expectedSigHex));

      expect(
          Ed25519.verifyDetached(sodium, kp.publicKey, signedBytes, sig),
          isTrue);
    });

    test('verifyDetached rejects flipped bit', () {
      final authJson =
          _readJson('${fixturesDir.path}/auth/nonce-and-signature.json');
      final expectedSig = _fromHex(authJson['expected_sig_hex'] as String);
      final signedBytes =
          _fromHex(authJson['expected_signed_bytes_hex'] as String);
      final pubKey =
          _fromHex(authJson['signer_public_key_hex'] as String);

      final tampered = Uint8List.fromList(expectedSig);
      tampered[0] ^= 0x01;

      expect(
          Ed25519.verifyDetached(sodium, pubKey, signedBytes, tampered),
          isFalse);
    });
  });

  // ---------------------------------------------------------------------------
  // SHA-256
  // ---------------------------------------------------------------------------

  group('Sha256Vectors', () {
    test('empty string hash', () {
      final hash = Sha256.hash(Uint8List(0));
      expect(_toHex(hash),
          equals(
              'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855'));
    });

    test('"abc" hash', () {
      final hash = Sha256.hash(utf8.encode('abc') as Uint8List);
      expect(_toHex(hash),
          equals(
              'ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad'));
    });

    test('ciphertext hash matches fixture', () {
      final cbor = CanonicalCborMapper();
      final envelopeBytes = File(
              '${fixturesDir.path}/envelopes/basic-1-recipient.cbor')
          .readAsBytesSync();
      final envelope = cbor.readValue(envelopeBytes);

      final ciphertext = envelope['ciphertext'] as Uint8List;
      final ciphertextHash = envelope['ciphertext_hash'] as Uint8List;

      final computed = Sha256.hash(ciphertext);
      expect(computed, equals(ciphertextHash));
    });
  });

  // ---------------------------------------------------------------------------
  // UUIDv7
  // ---------------------------------------------------------------------------

  group('UuidV7Vectors', () {
    test('all fixture UUIDs match', () {
      final rows =
          jsonDecode(File('${fixturesDir.path}/inputs/uuids.expected.json')
              .readAsStringSync()) as List<dynamic>;

      for (final row in rows) {
        final name = row['name'] as String;
        final tsMs = row['ts_ms'] as int;
        final fullRand = _fromHex(row['rand_hex'] as String);
        final rand10 = Uint8List.fromList(fullRand.sublist(0, 10));
        final expectedHex = row['expected_hex'] as String;

        final uuidBytes = UuidV7.generate(tsMs, rand10);
        expect(_toHex(uuidBytes), equals(expectedHex),
            reason: 'UUIDv7 for $name');
      }
    });

    test('generated UUID has version 7 and RFC variant', () {
      final bytes = UuidV7.now(sodium);
      // Version nibble is at byte 6, high 4 bits
      final version = (bytes[6] >> 4) & 0x0F;
      expect(version, equals(7));
      // Variant bits at byte 8: must be 10xxxxxx
      final variant = (bytes[8] >> 6) & 0x03;
      expect(variant, equals(2));
    });
  });

  // ---------------------------------------------------------------------------
  // Fingerprint
  // ---------------------------------------------------------------------------

  group('FingerprintVectors', () {
    test('all fixture fingerprints match', () {
      final rows =
          jsonDecode(File('${fixturesDir.path}/fingerprints/pubkey-to-fp.json')
              .readAsStringSync()) as List<dynamic>;

      for (final row in rows) {
        final name = row['name'] as String;
        final pubKey = _fromHex(row['public_key_hex'] as String);
        final expectedRender = row['expected_fp_render'] as String;

        final actual = Fingerprint.render(pubKey);
        expect(actual, equals(expectedRender),
            reason: 'Fingerprint for $name');
      }
    });
  });
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

Uint8List _fromHex(String hex) {
  final result = Uint8List(hex.length ~/ 2);
  for (var i = 0; i < result.length; i++) {
    result[i] = int.parse(hex.substring(i * 2, i * 2 + 2), radix: 16);
  }
  return result;
}

String _toHex(Uint8List bytes) {
  return bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
}

Map<String, dynamic> _readJson(String path) {
  return jsonDecode(File(path).readAsStringSync()) as Map<String, dynamic>;
}

File: src/test/java/com/wilhelmsen/cbslink/plugin/datawallet/crypto/PrimitivesFixtureTest.java
package com.wilhelmsen.cbslink.plugin.datawallet.crypto;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.HexFormat;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.assertj.core.api.Assertions.fail;

class PrimitivesFixtureTest {

    private static final ObjectMapper JSON = new ObjectMapper();
    private static final HexFormat HEX = HexFormat.of();
    private static Path fixturesDir;

    @BeforeAll
    static void resolveFixtures() {
        fixturesDir = resolveFixturesDir();
    }

    @Nested
    class KekDerivation {

        @Test
        void allPasswordFixturesProduceExpectedKek() throws IOException {
            List<JsonNode> expected = JSON.readValue(
                    fixturesDir.resolve("inputs/passwords.expected.json").toFile(),
                    new TypeReference<>() {}
            );
            for (JsonNode row : expected) {
                String name = row.get("name").asText();
                String password = row.get("password").asText();
                byte[] salt = HEX.parseHex(row.get("kdf_salt_hex").asText());
                JsonNode params = row.get("kdf_params");
                long m = params.get("m").asLong();
                long t = params.get("t").asLong();
                int p = params.get("p").asInt();
                String expectedKekHex = row.get("expected_kek_hex").asText();

                byte[] kek = Argon2id.deriveKek(password, salt, m, t, p);

                assertThat(HEX.formatHex(kek))
                        .as("KEK for %s", name)
                        .isEqualTo(expectedKekHex);
            }
        }

        @Test
        void rejectsParallelismNotOne() {
            byte[] salt = new byte[16];
            assertThatThrownBy(() -> Argon2id.deriveKek("test", salt, 65536, 1, 2))
                    .isInstanceOf(CryptoException.class)
                    .hasMessageContaining("p=1");
        }
    }

    @Nested
    class SecretBoxRoundTrip {

        @Test
        void encryptDecryptRoundTrip() {
            byte[] key = Random.bytes(32);
            byte[] nonce = Random.bytes(24);
            byte[] plaintext = "hello secretbox".getBytes(StandardCharsets.UTF_8);

            byte[] ciphertext = SecretBox.seal(plaintext, nonce, key);
            byte[] recovered = SecretBox.open(ciphertext, nonce, key);

            assertThat(recovered).isEqualTo(plaintext);
        }

        @Test
        void wrongKeyFailsToDecrypt() {
            byte[] key = Random.bytes(32);
            byte[] wrongKey = Random.bytes(32);
            byte[] nonce = Random.bytes(24);
            byte[] plaintext = "test".getBytes(StandardCharsets.UTF_8);

            byte[] ciphertext = SecretBox.seal(plaintext, nonce, key);

            assertThatThrownBy(() -> SecretBox.open(ciphertext, nonce, wrongKey))
                    .isInstanceOf(CryptoException.class);
        }
    }

    @Nested
    class SealedBoxOpen {

        @Test
        void sealAndOpenRoundTrip() throws IOException {
            JsonNode keypairs = JSON.readTree(fixturesDir.resolve("inputs/keypairs.json").toFile());
            byte[] seed = HEX.parseHex(keypairs.get("alice_enc").get("seed_hex").asText());
            X25519.KeyPair kp = X25519.seedKeypair(seed);

            byte[] plaintext = "sealed box test".getBytes(StandardCharsets.UTF_8);
            byte[] ciphertext = SealedBox.seal(plaintext, kp.publicKey());
            byte[] recovered = SealedBox.open(ciphertext, kp.publicKey(), kp.secretKey());

            assertThat(recovered).isEqualTo(plaintext);
        }

        @Test
        void openWrappedDataKeyFromFixture() throws IOException {
            JsonNode keypairs = JSON.readTree(fixturesDir.resolve("inputs/keypairs.json").toFile());
            byte[] aliceSeed = HEX.parseHex(keypairs.get("alice_enc").get("seed_hex").asText());
            X25519.KeyPair aliceKp = X25519.seedKeypair(aliceSeed);

            JsonNode envelopeMeta = JSON.readTree(fixturesDir.resolve("envelopes/envelope_meta.json").toFile());
            JsonNode basicMeta = envelopeMeta.get("basic-1-recipient");
            byte[] expectedDataKey = HEX.parseHex(basicMeta.get("data_key_hex").asText());

            CanonicalCborMapper cbor = new CanonicalCborMapper();
            byte[] envelopeBytes = Files.readAllBytes(fixturesDir.resolve("envelopes/basic-1-recipient.cbor"));
            @SuppressWarnings("unchecked")
            Map<String, Object> envelope = cbor.readValue(envelopeBytes, Map.class);
            @SuppressWarnings("unchecked")
            List<Map<String, Object>> wrappings = (List<Map<String, Object>>) envelope.get("recipient_wrappings");
            byte[] wrappedDataKey = (byte[]) wrappings.getFirst().get("wrapped_data_key");

            byte[] dataKey = SealedBox.open(wrappedDataKey, aliceKp.publicKey(), aliceKp.secretKey());
            assertThat(dataKey).isEqualTo(expectedDataKey);
        }
    }

    @Nested
    class Ed25519Determinism {

        @Test
        void seedKeypairMatchesExpectedPublicKeys() throws IOException {
            JsonNode keypairs = JSON.readTree(fixturesDir.resolve("inputs/keypairs.json").toFile());
            JsonNode expected = JSON.readTree(fixturesDir.resolve("inputs/keypairs.expected.json").toFile());

            var it = keypairs.fields();
            while (it.hasNext()) {
                var entry = it.next();
                String name = entry.getKey();
                String type = entry.getValue().get("type").asText();
                if (!"ed25519".equals(type)) continue;

                byte[] seed = HEX.parseHex(entry.getValue().get("seed_hex").asText());
                Ed25519.KeyPair kp = Ed25519.seedKeypair(seed);
                String expectedPubHex = expected.get(name).get("public_key_hex").asText();

                assertThat(HEX.formatHex(kp.publicKey()))
                        .as("Ed25519 public key for %s", name)
                        .isEqualTo(expectedPubHex);
            }
        }

        @Test
        void x25519SeedKeypairMatchesExpectedPublicKeys() throws IOException {
            JsonNode keypairs = JSON.readTree(fixturesDir.resolve("inputs/keypairs.json").toFile());
            JsonNode expected = JSON.readTree(fixturesDir.resolve("inputs/keypairs.expected.json").toFile());

            var it = keypairs.fields();
            while (it.hasNext()) {
                var entry = it.next();
                String name = entry.getKey();
                String type = entry.getValue().get("type").asText();
                if (!"x25519".equals(type)) continue;

                byte[] seed = HEX.parseHex(entry.getValue().get("seed_hex").asText());
                X25519.KeyPair kp = X25519.seedKeypair(seed);
                String expectedPubHex = expected.get(name).get("public_key_hex").asText();

                assertThat(HEX.formatHex(kp.publicKey()))
                        .as("X25519 public key for %s", name)
                        .isEqualTo(expectedPubHex);
            }
        }

        @Test
        void authSignatureMatchesFixture() throws IOException {
            JsonNode authFixture = JSON.readTree(fixturesDir.resolve("auth/nonce-and-signature.json").toFile());
            JsonNode keypairs = JSON.readTree(fixturesDir.resolve("inputs/keypairs.json").toFile());

            String signerName = authFixture.get("signer").asText();
            byte[] seed = HEX.parseHex(keypairs.get(signerName).get("seed_hex").asText());
            Ed25519.KeyPair kp = Ed25519.seedKeypair(seed);

            byte[] nonce = HEX.parseHex(authFixture.get("nonce_hex").asText());
            byte[] prefix = HEX.parseHex(authFixture.get("prefix_hex").asText());
            byte[] expectedSignedBytes = HEX.parseHex(authFixture.get("expected_signed_bytes_hex").asText());
            byte[] expectedSig = HEX.parseHex(authFixture.get("expected_sig_hex").asText());

            byte[] signedBytes = new byte[prefix.length + nonce.length];
            System.arraycopy(prefix, 0, signedBytes, 0, prefix.length);
            System.arraycopy(nonce, 0, signedBytes, prefix.length, nonce.length);
            assertThat(signedBytes).isEqualTo(expectedSignedBytes);

            byte[] signature = Ed25519.signDetached(kp.privateKey(), signedBytes);
            assertThat(HEX.formatHex(signature))
                    .isEqualTo(authFixture.get("expected_sig_hex").asText());

            assertThat(Ed25519.verifyDetached(kp.publicKey(), signedBytes, signature)).isTrue();
        }

        @Test
        void verifyDetachedRejectsFlippedBit() throws IOException {
            JsonNode authFixture = JSON.readTree(fixturesDir.resolve("auth/nonce-and-signature.json").toFile());
            byte[] expectedSig = HEX.parseHex(authFixture.get("expected_sig_hex").asText());
            byte[] signedBytes = HEX.parseHex(authFixture.get("expected_signed_bytes_hex").asText());
            byte[] pubKey = HEX.parseHex(authFixture.get("signer_public_key_hex").asText());

            byte[] tamperedSig = Arrays.copyOf(expectedSig, expectedSig.length);
            tamperedSig[0] ^= 0x01;

            assertThat(Ed25519.verifyDetached(pubKey, signedBytes, tamperedSig)).isFalse();
        }
    }

    @Nested
    class Sha256Vectors {

        @Test
        void emptyStringHash() {
            byte[] hash = Sha256.hash(new byte[0]);
            assertThat(HEX.formatHex(hash))
                    .isEqualTo("e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855");
        }

        @Test
        void abcHash() {
            byte[] hash = Sha256.hash("abc".getBytes(StandardCharsets.UTF_8));
            assertThat(HEX.formatHex(hash))
                    .isEqualTo("ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad");
        }

        @Test
        void envelopeCiphertextHashMatchesFixture() throws IOException {
            CanonicalCborMapper cbor = new CanonicalCborMapper();
            byte[] envelopeBytes = Files.readAllBytes(fixturesDir.resolve("envelopes/basic-1-recipient.cbor"));
            @SuppressWarnings("unchecked")
            Map<String, Object> envelope = cbor.readValue(envelopeBytes, Map.class);

            byte[] ciphertext = (byte[]) envelope.get("ciphertext");
            byte[] ciphertextHash = (byte[]) envelope.get("ciphertext_hash");

            byte[] computed = Sha256.hash(ciphertext);
            assertThat(computed).isEqualTo(ciphertextHash);
        }
    }

    @Nested
    class UuidV7Vectors {

        @Test
        void allFixtureUuidsMatch() throws IOException {
            List<JsonNode> uuids = JSON.readValue(
                    fixturesDir.resolve("inputs/uuids.expected.json").toFile(),
                    new TypeReference<>() {}
            );
            for (JsonNode row : uuids) {
                String name = row.get("name").asText();
                long tsMs = row.get("ts_ms").asLong();
                byte[] fullRand = HEX.parseHex(row.get("rand_hex").asText());
                byte[] rand10 = Arrays.copyOf(fullRand, 10);
                String expectedHex = row.get("expected_hex").asText();

                UUID uuid = UuidV7.generate(tsMs, rand10);
                String actualHex = HEX.formatHex(uuidToBytes(uuid));

                assertThat(actualHex)
                        .as("UUIDv7 for %s", name)
                        .isEqualTo(expectedHex);
            }
        }

        @Test
        void generatedUuidHasVersion7AndRfcVariant() {
            UUID uuid = UuidV7.now();
            assertThat(uuid.version()).isEqualTo(7);
            assertThat(uuid.variant()).isEqualTo(2);
        }
    }

    @Nested
    class FingerprintVectors {

        @Test
        void allFixtureFingerprintsMatch() throws IOException {
            List<JsonNode> fingerprints = JSON.readValue(
                    fixturesDir.resolve("fingerprints/pubkey-to-fp.json").toFile(),
                    new TypeReference<>() {}
            );
            for (JsonNode row : fingerprints) {
                String name = row.get("name").asText();
                byte[] pubKey = HEX.parseHex(row.get("public_key_hex").asText());
                String expectedRender = row.get("expected_fp_render").asText();

                String actual = Fingerprint.render(pubKey);

                assertThat(actual)
                        .as("Fingerprint for %s", name)
                        .isEqualTo(expectedRender);
            }
        }
    }

    @Nested
    class NoDirectLazySodiumUsage {

        @Test
        void noCodeOutsideCryptoPackageUsesLazySodiumDirectly() throws IOException {
            Path srcMain = fixturesDir.getParent().getParent().resolve("src/main/java");
            if (!Files.isDirectory(srcMain)) {
                fail("Could not find src/main/java from fixtures dir");
            }
            try (var stream = Files.walk(srcMain)) {
                List<Path> violations = stream
                        .filter(p -> p.toString().endsWith(".java"))
                        .filter(p -> !p.toString().contains("/crypto/"))
                        .filter(p -> {
                            try {
                                String content = Files.readString(p);
                                return content.contains("LazySodium")
                                        || content.contains("SodiumJava")
                                        || content.contains("com.goterl.lazysodium");
                            } catch (IOException e) {
                                return false;
                            }
                        })
                        .toList();
                assertThat(violations)
                        .as("Files outside crypto/ that reference LazySodium directly")
                        .isEmpty();
            }
        }

        @Test
        void noSecureRandomInMainSources() throws IOException {
            Path srcMain = fixturesDir.getParent().getParent().resolve("src/main/java");
            if (!Files.isDirectory(srcMain)) {
                fail("Could not find src/main/java from fixtures dir");
            }
            try (var stream = Files.walk(srcMain)) {
                List<Path> violations = stream
                        .filter(p -> p.toString().endsWith(".java"))
                        .filter(p -> {
                            try {
                                return Files.readString(p).contains("SecureRandom");
                            } catch (IOException e) {
                                return false;
                            }
                        })
                        .toList();
                assertThat(violations)
                        .as("Files that reference SecureRandom")
                        .isEmpty();
            }
        }
    }

    private static byte[] uuidToBytes(UUID uuid) {
        byte[] bytes = new byte[16];
        long msb = uuid.getMostSignificantBits();
        long lsb = uuid.getLeastSignificantBits();
        for (int i = 0; i < 8; i++) {
            bytes[i] = (byte) (msb >>> (56 - i * 8));
            bytes[8 + i] = (byte) (lsb >>> (56 - i * 8));
        }
        return bytes;
    }

    private static Path resolveFixturesDir() {
        Path dir = Paths.get(System.getProperty("user.dir"));
        for (int i = 0; i < 5; i++) {
            Path candidate = dir.resolve("spec").resolve("fixtures");
            if (Files.isDirectory(candidate) && Files.exists(candidate.resolve("manifest.json"))) {
                return candidate;
            }
            dir = dir.getParent();
            if (dir == null) break;
        }
        return fail("Could not find spec/fixtures directory from " + System.getProperty("user.dir"));
    }
}
